src/main/kotlin/no/skatteetaten/aurora/boober/feature/azure/AuroraAzureApimSubPart.kt

```diff
@@ -0,0 +1,88 @@
+import no.skatteetaten.aurora.boober.feature.namespace
+import no.skatteetaten.aurora.boober.model.AuroraDeploymentSpec
+import no.skatteetaten.aurora.boober.model.AuroraResource
+import no.skatteetaten.aurora.boober.model.openshift.ApimSpec
+import no.skatteetaten.aurora.boober.model.openshift.AuroraApim
+import no.skatteetaten.aurora.boober.service.AuroraDeploymentSpecValidationException
+
+val AuroraDeploymentSpec.isApimEnabled: Boolean
+    get() {
+        return this.getOrNull(AuroraAzureApimSubPart.ConfigPath.enabled) ?: false
+    }
+
+val AuroraDeploymentSpec.azureApimPolicies: List<String>?
+    get() {
+        //  Somewhat strange construction in order to avoid error with null as typeCast
+        val rawGroup = this.getOrNull<Any>(AuroraAzureApimSubPart.ConfigPath.policies) ?: return null
+        return rawGroup as List<String>
+    }
+
+class AuroraAzureApimSubPart {
+    object ConfigPath {
+        private const val root = "azure/apim"
+        const val enabled = "$root/enabled"
+        const val path = "$root/path"
+        const val openapiUrl = "$root/openapiUrl"
+        const val serviceUrl = "$root/serviceUrl"
+        const val policies = "$root/policies"
+        const val apiHost = "$root/apiHost"
+    }
+
+    fun generate(adc: AuroraDeploymentSpec, azureFeature: AzureFeature): Set<AuroraResource> {
+        return if (adc.isApimEnabled) {
+            val path: String = adc[ConfigPath.path]
+            val openapiUrl: String = adc[ConfigPath.openapiUrl]
+            val serviceUrl: String = adc[ConfigPath.serviceUrl]
+            val apiHost: String = adc[ConfigPath.apiHost]
+
+            setOf(
+                azureFeature.generateResource(
+                    AuroraApim(
+                        _metadata = newObjectMeta {
+                            name = adc.name
+                            namespace = adc.namespace
+                        },
+                        spec = ApimSpec(
+                            path = path,
+                            openapiUrl = openapiUrl,
+                            serviceUrl = serviceUrl,
+                            policies = adc.azureApimPolicies!!,
+                            apiHost = apiHost
+                        )
+                    )
+                )
+            )
+        } else {
+            emptySet()
+        }
+    }
+
+    fun validate(
+        adc: AuroraDeploymentSpec
+    ): List<Exception> {
+        val errors = mutableListOf<Exception>()
+        if (adc.isApimEnabled) {
+            Stream.of(
+                ConfigPath.path,
+                ConfigPath.openapiUrl,
+                ConfigPath.serviceUrl,
+                ConfigPath.apiHost
+            ).forEach {
+                if (adc.getOrNull<String>(it) == null) {
+                    errors.add(AuroraDeploymentSpecValidationException("You need to configure $it"))
+                }
+            }
+
+            if (adc.azureApimPolicies == null) {
```

Comment:
Sjekke for empty list? Eller er det dekket over? Synes uansett det er lurt, i tilfelle noen endrer litt et annet sted.

Reply:
Forsåvidt et godt poeng, men jeg ser at da jeg kodet dette tenkte jeg at det skulle være lov å ha tom liste, ref: https://github.com/Skatteetaten/boober/blob/feature/OTP-69-Support-AuroraApim-CRD/src/test/kotlin/no/skatteetaten/aurora/boober/feature/AuroraAzureApimSubPartPartTest.kt#L47

Jeg tenker at tanken er at du får lov til å være noe mer åpen i auroraconfig, og så har du en tiltagende hardere validering etterhvert som du går gjennom mudwell / mudbunny. Altså at det påkreves å ha et forhold til at du skal ha policies, men ikke nøyaktig hva det skal være. Det kan jo tenkes at mudbunny har lyst til å sette et default policy-sett dersom intet har vært angitt.

Analogt kan du merke at for URLene tester vi bare at de er gyldig formet, ikke at de faktisk har valid innhold.